[FLINK-12284][Network,Metrics]Fix the incorrect inputBufferUsage metric in credit-based network mode

Aitozi opened the pull request:
## What is the purpose of the change

This PR is to fix the bug in the calculation of  inputBufferUsage. It does't now take the exclusive buffer which is assigned from global buffer pool into account. 


## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*


## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

Aitozi commented:
Hi @zhijiangW could you help take a look on this PR when you are free, thanks.

Aitozi commented:
Thanks @zhijiangW for sharing your thoughts on this issue.
Although the semantics  of `inputBufferUsage` means the buffer occupied by `RemoteChannel` which may fill with data or not as you mentioned in credit-based mode in the moment,  But I think this PR  taking the exclusive buffer into account, which will make the semantics of this more close to the original intention.
I'am also ok to stay the same until the time of introducing new network related metrics. I think we are short of more detail metric about credit-based network model too.
Thanks for remind of the PR title format, I will pay more attention to this.:)

zhijiangW commented:
@pnowojski What do you think of this issue?

pnowojski commented:
> The current inputBufferUsage is only reflecting the usage of floating buffers, and it could give an indicator whether the setting of floating-buffers-per-gate is enough for performance tuning.

I agree, there is a value of the current semantic. 

I would even build upon that. If you want to detect bottleneck, current semantic of tracking just floating buffers is perfect for that. If current operator (reading from this input) is causing a bottleneck it well can be that only a single one input channel is back-pressured. At the moment we could detect this situation with "all floating buffers are exhausted/used".

> We could always assume if the exclusive buffers are enough for one channel, it will not request floating buffers. In other words, if we see the floating buffers are used, that means the exclusive buffers should also be used.

This is not true for even slight data skew.

I think that there is a value of tracking separately floating and exclusive buffers. 

Maybe we should make `inPoolUsage` track `floating + exclusive` buffers (it would make it consistent with non credit based mode), while we could add `inFloatingPoolUsage` metric with current semantic? What do you think?

Aitozi commented:
Hi @pnowojski , what is the reason for the second point ? 
> This is not true for even slight data skew

And +1 for adding the `inFloatingPoolUsage` specialized for detecting bottleneck.

zhijiangW commented:
Thanks for the confirmation and good suggestions. @pnowojski 

> > In other words, if we see the floating buffers are used, that means the exclusive buffers should also be used.
> > This is not true for even slight data skew

Maybe my previous expression is not very clear. If floating buffers are used, I only mean the exclusive buffers for one/some `RemoteInputChannels` are also expected to be used eventually, but not indicate the exclusive buffers for all the channels are used. 

For example, if the producer's backlog is 1, we would always request another 1 floating buffer even though the 2 exclusive buffers for this channel are available atm. Because we want to feedback some overhead credits beforehand in order not to block the network transport after producing more backlog soon. So it is not strong consistent for exclusive buffers used in time, might be eventual consistent within our expectation. If the backlog is becoming 0 from 1, the previous requested floating buffer would also be released by this channel if the 2 exclusive buffers are still available. So from the aspect of one input channel, it would not occupy extra floating buffers if its available exclusive buffers are enough.

I also agree with the above suggestions for distinguishing the total `inPoolUsage` and `floatingInPoolUsage`, or we could only retain `exclusiveInPoolUsage` and `floatingInPoolUsage` for credit-based mode.

Aitozi commented:
Hi, @zhijiangW thanks for your detail explanation, I make a conclusion below 
Non-credit-based mode:
- InPoolUsage

Credit-based mode:
- exclusiveInPoolUsage  exclusiveBufferUsed/exclusiveBufferTotal(all initial credit)
- floatingInPoolUsage   floatingBufferUsed/floatingBufferTotal
- InPoolUsage  combine with exclusiveInPoolUsage and floatingInPoolUsage

I think this looks better to distinguish the buffer usage between floating and exclusive in credit-based mode and also give a view of the overall perspective with inPoolUsage metric.  WDYT @pnowojski @zhijiangW

pnowojski commented:
@Aitozi 

I think your proposal makes sense. Having 3 metrics for credit based flow control sounds as the best, the most explicit solution for me.

> Maybe my previous expression is not very clear. If floating buffers are used, I only mean the exclusive buffers for one/some RemoteInputChannels are also expected to be used eventually, but not indicate the exclusive buffers for all the channels are used.

👍 

> Hi @pnowojski , what is the reason for the second point ?

Which second point do you mean @Aitozi ?

Aitozi commented:
OK, I will open another issue additionally to track and will update this PR accordingly, thanks.

Aitozi commented:
Have added the three metric for the credit based mode, please take a look when you are free @zhijiangW @pnowojski

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolMetric.java:
```diff
@@ -0,0 +1,136 @@
+	/**
+	 * Gauge metric measuring the floating buffer usage for {@link SingleInputGate}.
+	 */
+	public static class FloatingBufferPoolUsage implements Gauge<Float> {
```
Could we deduplicate the code here a bit? For example we could introduce `InputPoolUsageGauge` that would contain the following code/loop:
```
    protected abstract int calculateUsedBuffers(InputGate);

    protected abstract int calculateBufferPoolSize(InputGate);

	@Override
	public Float getValue() {
		int usedBuffers = 0;
		int bufferPoolSize = 0;

		for (SingleInputGate inputGate : inputGates) {
			usedBuffers += calculateUsedBuffers(inputGate)
			bufferPoolSize += calculateBufferPoolSize(inputGate);
		}

		if (bufferPoolSize != 0) {
			return ((float) usedBuffers) / bufferPoolSize;
		} else {
			return 0.0f;
		}
	}
```
That could be extended/implemented by `InputBufferPoolUsageGauge` and the classes that you introduced.

Also it would be nice to express somehow the total usage as `FloatingBufferPoolUsage` + `ExclusiveBufferPoolUsage` instead of duplicating the logic.

Aitozi replied:
It's a good suggestion, I will follow it.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -439,6 +439,14 @@ public int getInitialCredit() {
 		return initialCredit;
 	}
 
+	public int unsafeGetExclusiveBufferUsed() {
```
`unsynchronized` instead of `unsafe` to match existing `unsynchronizedGetNumberOfQueuedBuffers`?

zhijiangW replied:
ExclusiveBuffers

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolMetric.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Gauge metric measuring the input buffer pool usage gauge for {@link SingleInputGate}s under credit based mode.
+ */
+public class CreditBasedInputBufferPoolMetric {
```
Can you add tests for those classes?

Aitozi replied:
ok

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -439,6 +439,14 @@ public int getInitialCredit() {
+		return Math.max(0, initialCredit - bufferQueue.exclusiveBuffers.size());
+	}
+
+	public int unsafeGetFloatingBufferLeft() {
```
FloatingBuffers
Left->Used to consistent with above ExclusiveBuffersUsed?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolMetric.java:
```diff
@@ -0,0 +1,136 @@
+			if (requestedFloatingBuffer == 0) {
+				return 0.0f;
+			} else {
+				return 1 - ((float) leftFloatingBuffer) / requestedFloatingBuffer;
```
I think it is not very accurate here. 
It should be `(requestedFloatingBuffer - leftFloatingBuffer) / LocalBufferPool.getNumBuffers`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolMetric.java:
```diff
@@ -0,0 +1,136 @@
+		}
+
+		@Override
+		public Float getValue() {
```
We might extract the logics of above `exclusiveBuffersUsage and `floatingBuffersUsage into separate methods, then these methods could be reused here to get final results directly, to avoid below detail logics again.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -328,7 +332,17 @@ private void registerInputMetrics(MetricGroup inputGroup, MetricGroup buffersGro
 		buffersGroup.gauge(METRIC_INPUT_QUEUE_LENGTH, new InputBuffersGauge(inputGates));
-		buffersGroup.gauge(METRIC_INPUT_POOL_USAGE, new InputBufferPoolUsageGauge(inputGates));
+
+		if (config.isCreditBased()) {
```
For exclusive case there are no buffer pool actually. So maybe we could break down into `ExclusiveBuffersUsageGauge`, `FloatingBuffersUsageGauge`, `InputBuffersUsageGauge` instead?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -75,6 +76,9 @@
 	private static final String METRIC_OUTPUT_POOL_USAGE = "outPoolUsage";
 	private static final String METRIC_INPUT_QUEUE_LENGTH = "inputQueueLength";
 	private static final String METRIC_INPUT_POOL_USAGE = "inPoolUsage";
+	private static final String METRIC_CREDIT_BASED_FLOATING_BUFFER_POOL_USAGE = "floatingBufferUsage";
```
I think it is better to not add the keywords `CREDIT-BASED` here, because after we remove the deprecated network codes finally, it is no need to emphasize this term any more. The terms of exclusive/floating already indicate that.
Also it should be consistent between variable name and value, like `METRIC_FLOATING_BUFFERS_USAGE = "floatingBuffersUsage"`

zhijiangW commented:
Thanks for the updates @Aitozi ! I left some inline comments.

Aitozi commented:
Hi @zhijiangW @pnowojski I have addressed your comments. About the `inPoolUsage` name, I use the name for both two mode to be consist with previous style. And i squash the commit history solving the conflicts, please help review when you are free, I will continue to solve another PR after this one has been approved to avoid conflicts.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,207 @@
+			remoteInputChannel1.onSenderBacklog(3);
+			assertEquals(buffersPerChannel + 3, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
+
+			for (int i = 0; i < buffersPerChannel + 3; i++) {
```
nit: maybe deduplicate this loop into a separate method?

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,207 @@
+			for (Buffer buffer : exclusiveBuffers) {
+				segments.add(buffer.getMemorySegment());
+			}
+			network.getNetworkBufferPool().recycleMemorySegments(segments);
```
Couple of questions:
1. Shouldn't this be a part of `finally`?
2. Couldn't you call `buffer.recycle()`?
3. Maybe it would be better to run whole test with a `ClosableRegistry` that you would be populating as you create more objects? For example input gates there once you create them, or register `() -> { buffer.recycle() }` every time you just requested a buffer? It would make closing/releasing resource more reliable and more fine grained.

Aitozi commented:
Sorry for the late response @pnowojski , I will solve the comments you mentioned.

Aitozi commented:
Hi @pnowojski , I have adjust the test case according to your suggestion, please take a look when you have time, Thanks.

Aitozi commented:
ping @pnowojski @zhijiangW

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -117,80 +118,61 @@ public void testBufferUsageMetrics() throws IOException, InterruptedException {
-				segments.add(buffer.getMemorySegment());
-			}
+
 			network.getNetworkBufferPool().recycleMemorySegments(segments);
```
This I think is a left over - dead code - `segments` seemts to be always empty.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -43,19 +45,18 @@
  */
 public class InputBuffersMetricsTest {
 
+	private CloseableRegistry closeableRegistry = new CloseableRegistry();
```
one new issue: you can not close the same `CloseableRegistry` more then once, so you would have to set it up/create new instance in `@org.junit.Before` method. 

nit: you could also close it in the `@After` method.

Aitozi replied:
I think the `closeableRegistry` do not have to share between test cases, so i move this to the method scope.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -117,80 +118,61 @@ public void testBufferUsageMetrics() throws IOException, InterruptedException {
 			inputGate2.close();
-			network.shutdown();
+			closeableRegistry.close();
+			network.close();
```
nit: you could add `network.close()` and `inputGateN.close()` to the `closeableRegistry`

Aitozi replied:
I think this is not so good, because the closeableRegistry do not close respect to the register order which may lead to network closed before network buffers has all be recycled.

Aitozi commented:
Thanks @pnowojski for you advice, I have addressed some of them, please have a look again 😄

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -84,6 +87,9 @@
 	private static final String METRIC_INPUT_POOL_USAGE = "inPoolUsage";
+	private static final String METRIC_INPUT_FLOATING_BUFFERS_USAGE = "floatingBuffersUsage";
+	private static final String METRIC_INPUT_EXCLUSIVE_BUFFERS_USAGE = "exclusiveBuffersUsage";
+
```
remove this extra empty line

Aitozi replied:
ok，the checkstyle seems not cover this, so i overlooked this.

zhijiangW replied:
it is not covered by checkstyle, but I think it should be common sense. :)

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -303,7 +309,18 @@ private void registerInputMetrics(MetricGroup inputGroup, MetricGroup buffersGro
+			buffersGroup.gauge(METRIC_INPUT_EXCLUSIVE_BUFFERS_USAGE, exclusiveBuffersUsageGauge);
+			buffersGroup.gauge(METRIC_INPUT_FLOATING_BUFFERS_USAGE, floatingBuffersUsageGauge);
+			buffersGroup.gauge(METRIC_INPUT_POOL_USAGE, new CreditBasedInputBufferPoolUsageGauge(floatingBuffersUsageGauge,
+				exclusiveBuffersUsageGauge, inputGates));
```
If the parameters are broken into separate lines, make every parameter in a separate line.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolUsageGauge.java:
```diff
@@ -0,0 +1,48 @@
+	public CreditBasedInputBufferPoolUsageGauge(FloatingBuffersUsageGauge floatingBuffersUsageGauge,
+												ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge,
+												SingleInputGate[] inputGates) {
+		super(inputGates);
```
`checkNotNull` for the parameters

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/ExclusiveBuffersUsageGauge.java:
```diff
@@ -0,0 +1,55 @@
+public class ExclusiveBuffersUsageGauge extends AbstractBuffersUsageGauge {
+
+	public ExclusiveBuffersUsageGauge(SingleInputGate[] inputGates) {
+		super(inputGates);
```
ditto: `checkNotNull(inputGates)`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/FloatingBuffersUsageGauge.java:
```diff
@@ -0,0 +1,60 @@
+public class FloatingBuffersUsageGauge extends AbstractBuffersUsageGauge {
+
+	public FloatingBuffersUsageGauge(SingleInputGate[] inputGates) {
+		super(inputGates);
```
ditto: `checkNotNull`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/FloatingBuffersUsageGauge.java:
```diff
@@ -0,0 +1,60 @@
+
+/**
+ * Gauge metric measuring the floating buffers usage gauge for {@link SingleInputGate}s.
+ * real used buffers which are occupied by intermediate data / localBufferPool size
```
`real used buffers which are occupied by intermediate data / localBufferPool size` seems not very proper, might be removed.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/AbstractBuffersUsageGauge.java:
```diff
@@ -0,0 +1,58 @@
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate;
+
+/**
+ * Abstract gauge implementation for calculate the buffer usage percent.
```
for calculating the buffers

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/AbstractBuffersUsageGauge.java:
```diff
@@ -0,0 +1,58 @@
+	public abstract int calculateUsedBuffers(SingleInputGate inputGate);
+
+	@VisibleForTesting
+	public abstract int calculateBufferPoolSize(SingleInputGate inputGate);
```
how about renaming `calculateBufferPoolSize` as `calculateTotalBuffers`, because for exclusive case it has not buffer pool, so this naming seems not general.

Aitozi replied:
agree

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/ExclusiveBuffersUsageGauge.java:
```diff
@@ -0,0 +1,55 @@
+
+	@Override
+	public int calculateBufferPoolSize(SingleInputGate inputGate) {
+		int requestedExclusiveBuffers = 0;
```
maybe rename as totalExclusiveBuffers

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -84,6 +87,9 @@
 	private static final String METRIC_OUTPUT_POOL_USAGE = "outPoolUsage";
 	private static final String METRIC_INPUT_QUEUE_LENGTH = "inputQueueLength";
 	private static final String METRIC_INPUT_POOL_USAGE = "inPoolUsage";
+	private static final String METRIC_INPUT_FLOATING_BUFFERS_USAGE = "floatingBuffersUsage";
```
might add `input` prefix as `inputFloatingBuffersUsage` and `inputExclusiveBuffersUsage`.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolUsageGauge.java:
```diff
@@ -0,0 +1,48 @@
+	private FloatingBuffersUsageGauge floatingBuffersUsageGauge;
+	private ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge;
+
+	public CreditBasedInputBufferPoolUsageGauge(FloatingBuffersUsageGauge floatingBuffersUsageGauge,
```
parameters formatting : should be separate line with indentation for every parameter

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolUsageGauge.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class CreditBasedInputBufferPoolUsageGauge extends AbstractBuffersUsageGauge {
+
+	private FloatingBuffersUsageGauge floatingBuffersUsageGauge;
```
final

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolUsageGauge.java:
```diff
@@ -0,0 +1,48 @@
+public class CreditBasedInputBufferPoolUsageGauge extends AbstractBuffersUsageGauge {
+
+	private FloatingBuffersUsageGauge floatingBuffersUsageGauge;
+	private ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge;
```
ditto: final

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/AbstractBuffersUsageGauge.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public abstract class AbstractBuffersUsageGauge implements Gauge<Float> {
+
+	protected SingleInputGate[] inputGates;
```
final

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolUsageGauge.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Gauge metric measuring the input buffer pool usage gauge for {@link SingleInputGate}s under credit based mode.
+ */
+public class CreditBasedInputBufferPoolUsageGauge extends AbstractBuffersUsageGauge {
```
This class should be rename as `CreditBasedInputBuffersUsageGauge`, because the exclusive buffers have no pool and it would also be covered in this gauge.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -303,7 +309,18 @@ private void registerInputMetrics(MetricGroup inputGroup, MetricGroup buffersGro
 		buffersGroup.gauge(METRIC_INPUT_QUEUE_LENGTH, new InputBuffersGauge(inputGates));
-		buffersGroup.gauge(METRIC_INPUT_POOL_USAGE, new InputBufferPoolUsageGauge(inputGates));
+
+		if (config.isCreditBased()) {
```
Note: in another PR #8485 the metrics operations are refactored in a metric factory. If that PR merged firstly then we might rebase to adjust this part.

Aitozi replied:
ok

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/AbstractBuffersUsageGauge.java:
```diff
@@ -0,0 +1,58 @@
+		int usedBuffers = 0;
+		int bufferPoolSize = 0;
+
+		for (SingleInputGate inputGate : inputGates) {
```
I ever had a concern with the case of multiple input gates. Each `SingleInputGate` corresponds to a different upstream `JobVertex`. If we calculate the value for all the gates, it seems difficult to distinguish the specific usage for individual gate for performance tuning.

Another option is we calculate the value separately for each gate, and take the maximum value as the final value if we agree the max usages should bring more concerns for performance tuning. I am not quite sure of it. WDYT?@pnowojski

Aitozi replied:
My opinion here: although we calculated the maximum value we seems can't not know the reported value comes from which inputGate? Unless we report the individual gate metric with gate specific on metric name, but it seems too detailed with too much metric data.

zhijiangW replied:
Yes, I only pointed out a different idea and have not thought through it yet. 
The current way keeps the previous behavior and I could accept it atm if no better solution.

Aitozi replied:
Agree with you that mixing serval inputGate usage will make the performance tuning harder, looking forward to better solution , I will also keep an eye on this, too :)

pnowojski replied:
I haven't thought about this. I would keep it as it is in order for this to be merged before 1.9 feature freeze and this has already grown substantially from its initial form. There can be some follow up work in the future.

zhijiangW replied:
agree

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/InputBufferPoolUsageGauge.java:
```diff
@@ -18,38 +18,33 @@
 
 	public InputBufferPoolUsageGauge(SingleInputGate[] inputGates) {
-		this.inputGates = inputGates;
+		super(inputGates);
```
`checkNotNull`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+/**
+ * Test for input buffer usage related metrics.
+ */
+public class InputBuffersMetricsTest {
```
extends `TestLogger` for new added test class

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test for input buffer usage related metrics.
```
`Tests the metrics for input buffers usage`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+
+		CloseableRegistry closeableRegistry = new CloseableRegistry();
+
+		final NettyShuffleEnvironment network = new NettyShuffleEnvironmentBuilder().setIsCreditBased(true).build();
```
`setIsCreditBased(true)` could be ignored because the default value in builder is already true.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+	@Test
+	public void testBufferUsageMetrics() throws IOException, InterruptedException {
+
+		CloseableRegistry closeableRegistry = new CloseableRegistry();
```
If we want to use final for some vars, it is better to keep the same format for others in consistency.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+
+		final SingleInputGate inputGate1 = new SingleInputGateBuilder()
+			.setNumberOfChannels(2)
+			.setIsCreditBased(true)
```
remove `.setIsCreditBased(true)`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+
+		final SingleInputGate inputGate2 =  new SingleInputGateBuilder()
+			.setNumberOfChannels(2)
+			.setIsCreditBased(true)
```
ditto: `setIsCreditBased(true)`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+			final ConnectionID connectionId1 = new ConnectionID(new InetSocketAddress("localhost", 5000), 0);
+			final ConnectionID connectionId2 = new ConnectionID(new InetSocketAddress("localhost", 5000), 0);
+
+			RemoteInputChannel remoteInputChannel1 = createUnknownInputChannel(network, inputGate1, resultPartitionId1, 0).toRemoteInputChannel(connectionId1);
```
We might use `new InputChannelBuilder().buildRemoteAndSetToGate(inputGate1)` instead, also for the following local channel.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+			ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+			CreditBasedInputBufferPoolUsageGauge inputBufferPoolUsageGauge = new CreditBasedInputBufferPoolUsageGauge(floatingBuffersUsageGauge, exclusiveBuffersUsageGauge, inputGates);
+
+			inputMetric.gauge("floatingBuffersUsage", floatingBuffersUsageGauge);
```
`inputMetric.gauge` seems not unnecessary, also the `inputMetrics` definition also not necessary.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+			assertEquals(0.0, exclusiveBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, inputBufferPoolUsageGauge.getValue(), 0.0);
+
+			assertEquals(buffersPerChannel * 2, exclusiveBuffersUsageGauge.calculateBufferPoolSize(inputGate1));
```
the magic number 2 seems not easy to understand here, it is better to name a local var as `numberOfRemoteChannels=2`.
If we change the default `networkBuffersPerChannel` value in `NettyShuffleEnvironmentBuilder`, this test would fail, because the local var `buffersPerChannel` is not passed while creating `RemoteInputChannel`

zhijiangW replied:
Also for the case of following `extraNetworkBuffersPerGate`.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+
+			assertEquals(buffersPerChannel, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
+			assertEquals(1 / 3.0, exclusiveBuffersUsageGauge.getValue(), 0.0001);
+			assertEquals(buffersPerChannel / (extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0), inputBufferPoolUsageGauge.getValue(), 0.0001);
```
We could define a local var for `extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0` because it is used many times below.

zhijiangW replied:
Even we could extract a method for covering the `drainBuffer` with three asserts together if possible, because we do the drain logic three times which might reuse the same logic.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,180 @@
+			MetricGroup inputMetric = UnregisteredMetricGroups.createUnregisteredTaskMetricGroup().getIOMetricGroup().addGroup("buffers");
+
+			inputGate1.setup();
+			inputGate2.setup();
```
If possible we could extract a local method for extracting the relevant logics for creating `SingleInputGate`, including the InputGate/InputChannel creations and setup. Then it would not do the same things for two gates. The `InputChannel` used in following parts could also be got from created gate.

zhijiangW commented:
Thanks for the updates @Aitozi !

The abstract seems better now and I left some comments mainly for formatting and deduplications in tests.

Aitozi commented:
Thanks @zhijiangW for your careful review. About this comment

> covering the drainBuffer with three asserts together if possible

I think it's not so good, because the second and the third assert depend on the previous drainBuffer action means if add the assert in the drainBuffer, `drainBuffer` will be a function with state

Other comments has all been resolved, please have a look again, very thanks :)

Aitozi commented:
ping @zhijiangW @pnowojski , could you take a look again, thanks :)

Aitozi commented:
I will solve the conflicts.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBuffersUsageGauge.java:
```diff
@@ -0,0 +1,51 @@
+		ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge,
+		SingleInputGate[] inputGates) {
+		super(checkNotNull(inputGates));
+		this.floatingBuffersUsageGauge = floatingBuffersUsageGauge;
```
checkNotNull for `floatingBuffersUsageGauge` and `exclusiveBuffersUsageGauge`

zhijiangW commented:
@Aitozi thanks for addressing my comments.
Could you rebase again for solving the conflicts, then I could  double reviewing it.

zhijiangW commented:
In addition, you might need to execute the command "mvn package -Dgenerate-config-docs -DskipTests -pl flink-docs -am -nsu" to generate the updated html files.

Aitozi commented:
Hi, @zhijiangW I have solve the conflicts and generate the html, could you help review this again , thanks :)

zhijiangW commented on docs/_includes/generated/blob_server_configuration.html:
```diff
@@ -7,16 +7,16 @@
             <td>The connection timeout in milliseconds for the blob client.</td>
         </tr>
+        <tr>
+            <td><h5>blob.client.socket.timeout</h5></td>
```
Why this changed? I think it should not be changed, and you could double check.

Aitozi replied:
I also notice this, but i have not change this part, does this comes from some commit which does not generate the html？

zhijiangW replied:
I think it should not be that case. Maybe your metrics.md change would not effect any html files.

zhijiangW replied:
sorry for bringing you trouble here. I also found other person cause the same issue for unrelated html changes. I would double check the root PR who should cause this change, and you might ignore this change in your PR.

zhijiangW commented on docs/_includes/generated/netty_shuffle_environment_configuration.html:
```diff
@@ -0,0 +1,61 @@
+<table class="table table-bordered">
```
ditto: this file should not be changed as well.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBuffersUsageGauge.java:
```diff
@@ -0,0 +1,51 @@
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Gauge metric measuring the input buffer pool usage gauge for {@link SingleInputGate}s under credit based mode.
```
remove `pool`, `measuring the input buffers usage for `

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/NettyShuffleMetricFactory.java:
```diff
@@ -66,6 +66,8 @@
 
 	private static final String METRIC_INPUT_QUEUE_LENGTH = "inputQueueLength";
 	private static final String METRIC_INPUT_POOL_USAGE = "inPoolUsage";
+	private static final String METRIC_INPUT_FLOATING_BUFFERS_USAGE = "floatingBuffersUsage";
```
floatingBuffersUsage -> inputFloatingBuffersUsage, also add the `input` prefix for following `exclusiveBuffersUsage`, then it is easy to distinguish from the metric name and keep consistent with other.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/NettyShuffleMetricFactory.java:
```diff
@@ -154,24 +157,40 @@ private static void registerOutputMetrics(
+		} else {
+			buffersGroup.gauge(METRIC_INPUT_POOL_USAGE, new InputBufferPoolUsageGauge(inputGates));
+		}
 		buffersGroup.gauge(METRIC_INPUT_POOL_USAGE, new InputBufferPoolUsageGauge(inputGates));
```
should remove this, it was already done in above else.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/NettyShuffleMetricFactory.java:
```diff
@@ -154,24 +157,40 @@ private static void registerOutputMetrics(
+
+			buffersGroup.gauge(METRIC_INPUT_EXCLUSIVE_BUFFERS_USAGE, exclusiveBuffersUsageGauge);
+			buffersGroup.gauge(METRIC_INPUT_FLOATING_BUFFERS_USAGE, floatingBuffersUsageGauge);
+			buffersGroup.gauge(METRIC_INPUT_POOL_USAGE, new CreditBasedInputBuffersUsageGauge(floatingBuffersUsageGauge,
```
It is not good to make partial parameters in separate line. I think you should make `METRIC_INPUT_POOL_USAGE` as a separate line, and `new CreditBasedInputBuffersUsageGauge(floatingBuffersUsageGauge, exclusiveBuffersUsageGauge, inputGates)` as a separate line.
Or you could define a local var for `CreditBasedInputBuffersUsageGauge` like above `floatingBuffersUsageGauge` and `exclusiveBuffersUsageGauge`.

zhijiangW replied:
nit: I think it is better to define a local var for `CreditBasedInputBuffersUsageGauge` like above `floatingBuffersUsageGauge` and `exclusiveBuffersUsageGauge`, because the broken line seems not good here.

Aitozi replied:
ok

Aitozi replied:
@zhijiangW do you have other suggestions, I can solve them together today to merge this before 1.9 release

zhijiangW replied:
I think I have issues only for the tests, and I would post then later today.

Aitozi replied:
ok, very thanks

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+			.buildRemoteAndSetToGate(inputGate);
+	}
+
+	private LocalInputChannel buildLocalChannel(int channelIndex,
```
formatting: make every parameter in separate line and with a tab indentation.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+		}
+	}
+
+	private RemoteInputChannel buildRemoteChannel(int channelIndex,
```
formatting: make every parameter in separate line and with a tab indentation.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+			.buildLocalAndSetToGate(inputGate);
+	}
+
+	private Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> buildInputGate(NettyShuffleEnvironment network,
```
ditto: formatting

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+		int buffersPerChannel = 2;
+		int extraNetworkBuffersPerGate = 8;
+
+		final CloseableRegistry closeableRegistry = new CloseableRegistry();
```
you could remove final for these vars, otherwise all the vars should be final to keep consistent, like above int vars.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+			.setFloatingNetworkBuffersPerGate(extraNetworkBuffersPerGate)
+			.build();
+
+		final Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple1 = buildInputGate(network,
```
parameter formatting

zhijiangW replied:
Tuple is not a good choice here unless it is necessary to do so.

I think you only need to get one `RemoteInputChannel` for `requestSubpartition` and `onSenderBacklog` below. Another alternative way is you could construct `SingleInputGate` firstly, then construct relevant channel via `InputChannelBuilder#buildRemoteAndSetToGate(gate)`.

Aitozi replied:
I was meant to not build channel serval times to avoid duplicate code.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+		final Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple1 = buildInputGate(network,
+			numberOfRemoteChannelsGate1,
+			numberOfLocalChannelsGate1);
+		final Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple2 = buildInputGate(network,
```
ditto

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+			SingleInputGate[] inputGates = new SingleInputGate[]{tuple1.f0, tuple2.f0};
+			FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+			ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+			CreditBasedInputBuffersUsageGauge inputBufferPoolUsageGauge = new CreditBasedInputBuffersUsageGauge(floatingBuffersUsageGauge,
```
ditto

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+public class InputBuffersMetricsTest extends TestLogger {
+
+	@Test
+	public void testBufferUsageMetrics() throws IOException, InterruptedException {
```
TBH, I think this test is too big for mixing many functions to be verified together. 

My suggestion is breaking it into several simple ones to make sure each test is only for verifying one function, then it would be very easy for reviewing and maintaining. 

E.g. if someone changes the logic of internal backlog future which might cause this test fail, if he wants to fix this test, he has to review the whole more than 100 lines for tracing the reason. If each test is short and only covering single function, then it would be very easy for other person, also you could find it easy to write single unit test. I could give some examples below.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,203 @@
+			assertEquals(extraNetworkBuffersPerGate, floatingBuffersUsageGauge.calculateTotalBuffers(inputGate1));
+
+			assertEquals(buffersPerChannel * numberOfRemoteChannelsGate2, exclusiveBuffersUsageGauge.calculateTotalBuffers(inputGate2));
+			assertEquals(extraNetworkBuffersPerGate, floatingBuffersUsageGauge.calculateTotalBuffers(inputGate2));
```
If we write a single test only for verifying the function of `FloatingBuffersUsageGauge#calculateTotalBuffers`, then we only need one `SingleInputGate`, do not need to verify it twice for both `inputGate1` and `inputGate2`.

The array of `SingleInputGate` is only reasonable for verifying the function of `AbstractBuffersUsageGauge#getValue`.

Aitozi replied:
Yes, I will break this test case into serval independent methods.

zhijiangW commented:
In addition, one PR should not connect to two jira tickets in general. So I think you might need to close one jira and adjust this PR to only one jira.

pnowojski commented:
> In addition, one PR should not connect to two jira tickets in general. So I think you might need to close one jira and adjust this PR to only one jira.

I wouldn't be so strict about having couple of jira tickets in one PR, especially if they depend on one another. However as a rule of thumb one commit shouldn't be addressing two jira tickets, since probably it means that either:
- commit is fixing two separate things
- Jira tickets are duplicated

As far as I recall this is inline with our coding style guideline. Of course it's better to have separate things fixed in independent PRs.

Here I think it's the case that "Jira tickets are duplicated". In my opinion `Add metrics for floatingBufferUsage and exclusiveBufferUsage for credit based mode` is a solution for `InputBufferPoolUsage is incorrect in credit-based network control flow`. So I would close the newer ticket as duplicate of the older one.

Aitozi commented:
ok, I will fix this tonight

Aitozi commented:
Hi @zhijiangW , I've addressed your comments , please take a look again, thanks.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+		int numberOfRemoteChannels = 2;
+		int numberOfLocalChannels = 0;
+
+		int buffersPerChannel = 2;
```
buffersPerChannel  - > numberOfBufferPerChannel
extraNetworkBuffersPerGate -> numberOfBuffersPerGate

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+		closeableRegistry.registerCloseable(network::close);
+		closeableRegistry.registerCloseable(inputGate1::close);
+
+		try {
```
try (CloseableRegistry closeableRegistry = new CloseableRegistry()) to avoid finally clause.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+			SingleInputGate[] inputGates = new SingleInputGate[]{inputGate1};
+			FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+			ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+			CreditBasedInputBuffersUsageGauge inputBufferPoolUsageGauge = new CreditBasedInputBuffersUsageGauge(
```
All these constructors could be removed outside of `try`, and put above `closeableRegistry.registerCloseable` inside `try`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+		SingleInputGate inputGate2 = tuple2.f0;
+
+		try {
+			assertEquals(tuple1.f1.size(), numberOfRemoteChannelsGate1 + numberOfLocalChannelsGate1);
```
this assert is not very necessary here, because it is not brought by this change.

zhijiangW replied:
What do you think this issue?

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+			assertEquals((double) buffersPerChannel / (buffersPerChannel * numberOfRemoteChannelsTotal), exclusiveBuffersUsageGauge.getValue(), 0.0001);
+			assertEquals((double) buffersPerChannel / totalBuffers, inputBuffersUsageGauge.getValue(), 0.0001);
+
+			drainBuffer(buffersPerChannel, remoteInputChannel2, closeableRegistry);
```
It might be feasible to put `drainBuffer` and the following three asserts in a loop, because we do it twice.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+			assertEquals((double) buffersPerChannel * numberOfRemoteChannelsGate1 / totalBuffers,
+				inputBuffersUsageGauge.getValue(), 0.0001);
+		} finally {
+			inputGate1.close();
```
The relevant closes could also be added into `CloseableRegistry` as the first test done.

Aitozi replied:
As i mentioned above to @pnowojski , I think the closeableRegistry do not respect to the register order which will lead to some buffer have not be released with some excpetion

Aitozi replied:
And the closeableRegistry will close quietly, ignoring the exception (exclusive buffers may have not been released before close global buffer pool). So I prefer to keep the current shape.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+			assertEquals((double) (buffersPerChannel * numberOfRemoteChannelsGate1 + requestFloatingTotal) / totalBuffers,
+				inputBuffersUsageGauge.getValue(), 0.0001);
+		} finally {
+			inputGate1.close();
```
ditto: remove finally and add closes to `CloseableRegistry`.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+
+	private Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> buildInputGate(
+		NettyShuffleEnvironment network,
+		int remoteChannelNum,
```
remoteChannelNum -> numberOfRemoteChannels, also for the following local channel

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,301 @@
+
+		int channelIdx = 0;
+		for (int i = 0; i < remoteChannelNum; i++) {
+			res.f1.add(buildRemoteChannel(channelIdx, new ResultPartitionID(), inputGate, network));
```
We might not need to pass `new ResultPartitionID()` to the method, which can be done internally.
Move the methods of `buildRemoteChannel` and `buildLocalChannel` after this `buildInputGate` method.

zhijiangW commented:
Thanks for addressing my comments @Aitozi .
I only have some new comments for tests before merging. In addition I think you should adjust the PR and commit titles for covering only one jira.

Aitozi commented:
Ping @zhijiangW

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+	private Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> buildInputGate(
+		NettyShuffleEnvironment network,
+		int numberOfRemoteChannels,
+		int localChannelNum) throws IOException {
```
localChannelNum -> numberOfLocalChannels

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+			numberOfRemoteChannels,
+			numberOfLocalChannels);
+
+		SingleInputGate inputGate1 = tuple1.f0;
```
We only need `SingleInputGate` from tuple here, so it is simple `SinglerInputGate inputGate = buildInputGate(network, numberOfRemoteChannels, numberOfLocalChannels).f0` directly.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+		int numberOfLocalChannelsGate2 = 1;
+
+		int numberOfRemoteChannelsTotal = numberOfRemoteChannelsGate1 + numberOfRemoteChannelsGate2;
+		int numberOfInputGates = 2;
```
could remove `numberOfInputGates` and use `inputGates.length` in following `totalBuffers`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+		int numberOfRemoteChannelsGate2 = 1;
+		int numberOfLocalChannelsGate2 = 1;
+
+		int numberOfRemoteChannelsTotal = numberOfRemoteChannelsGate1 + numberOfRemoteChannelsGate2;
```
numberOfRemoteChannelsTotal -> totalNumberOfRemoteChannels

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+		try {
+			assertEquals(tuple1.f1.size(), numberOfRemoteChannelsGate1 + numberOfLocalChannelsGate1);
+
+			RemoteInputChannel remoteInputChannel1 = tuple1.f1.get(0);
```
We could move these initializations outside of `try`.
    ```
   RemoteInputChannel remoteInputChannel1 = tuple1.f1.get(0);
       RemoteInputChannel remoteInputChannel2 = tuple1.f1.get(1);
	
       SingleInputGate[] inputGates = new SingleInputGate[]{tuple1.f0, tuple2.f0};
	FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
	ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
	CreditBasedInputBuffersUsageGauge inputBuffersUsageGauge = new CreditBasedInputBuffersUsageGauge(
					floatingBuffersUsageGauge,
					exclusiveBuffersUsageGauge,
					inputGates);
```

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+		SingleInputGate inputGate2 = tuple2.f0;
+
+		try {
+			assertEquals(tuple1.f1.size(), numberOfRemoteChannelsGate1 + numberOfLocalChannelsGate1);
```
this assert is not necessary for metric changes , so we could remove it and then change to tuple2 instead.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+		SingleInputGate inputGate1 = tuple1.f0;
+		SingleInputGate inputGate2 = tuple2.f0;
+
+		try {
```
`try (CloseableRegistry closeableRegistry = new CloseableRegistry())` so in finally we could avoid closing it.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+			remoteInputChannel1.requestSubpartition(0);
+
+			int backlog = 3;
+			int requestFloatingTotal = buffersPerChannel + backlog;
```
requestFloatingTotal -> totalRequestedBuffers

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+			exclusiveBuffersUsageGauge,
+			inputGates);
+
+		try {
```
try (CloseableRegistry closeableRegistry = new CloseableRegistry())

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+		SingleInputGate inputGate2 = tuple2.f0;
+
+		RemoteInputChannel remoteInputChannel1 = tuple1.f1.get(0);
+		RemoteInputChannel remoteInputChannel2 = tuple1.f1.get(1);
```
It might use `List<RemoteInputChannel> remoteInputChannels = tuple1.f1` instead, then we could loop the channel to avoid `drainAndValidate` twice below.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+				inputBuffersUsageGauge,
+				inputGate1);
+
+			drainAndValidate(
```
we could merge two drains as below :
```
int channelIndex = 1;
for (RemoteInputChannel channel : remoteInputChannels) {
	drainAndValidate(
		buffersPerChannel,
		buffersPerChannel * channelIndex++,
		channel,
		closeableRegistry,
		totalBuffers,
		buffersPerChannel * numberOfRemoteChannelsTotal,
		exclusiveBuffersUsageGauge,
		inputBuffersUsageGauge,
		inputGate1);
}
```

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+		ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge,
+		CreditBasedInputBuffersUsageGauge inputBuffersUsageGauge,
+		SingleInputGate inputGate) throws IOException {
+		drainBuffer(numBuffersToRequest, channel, closeableRegistry);
```
better to add one empty line before `drainBuffer`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+			numberOfLocalChannelsGate2);
+
+		SingleInputGate inputGate1 = tuple1.f0;
+		SingleInputGate inputGate2 = tuple2.f0;
```
`SingleInputGate inputGate2 = buildInputGate(network, numberOfRemoteChannelsGate2, numberOfLocalChannelsGate2).f0` instead, we could avoid long var tuple2.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,316 @@
+
+			// drain gate1's exclusive buffers
+			drainBuffer(buffersPerChannel, remoteInputChannel1, closeableRegistry);
+			drainBuffer(buffersPerChannel, remoteInputChannel2, closeableRegistry);
```
we might only need drain `remoteInputChannel1`, `remoteInputChannel2` is actually not used following.

zhijiangW commented:
Thanks for the updates @Aitozi ! I left some other comments for tests.

Aitozi commented:
Very thanks for @zhijiangW 's carefully review, please take a look again, hoping this is the last look :)

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,293 @@
+		}
+	}
+
+	private Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> buildInputGate(
```
we could only return `Tuple2<SingleInputGate, List<RemoteInputChannel>>` instead because the list of local channels would not be used any more.

Aitozi replied:
yes

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java:
```diff
@@ -0,0 +1,293 @@
+			.buildRemoteAndSetToGate(inputGate);
+	}
+
+	private LocalInputChannel buildLocalChannel(
```
make it void method after adjusting tuple2 above.

zhijiangW commented:
Thanks for the patience and addressing my comments @Aitozi !

I think you might miss one previous comment for using tuple2 instead of tuple3. After fixing this I have no other concerns. 👍 LGTM!

Aitozi commented:
Ping @pnowojski

pnowojski commented:
Thanks for the fix @Aitozi and your patience with our reviews (especially that the scope of this ticket has grown significantly from your initial version). Merging as I can see that on your private travis this is already green.
